apps/app_voicemail.c

```diff
@@ -11413,6 +11412,371 @@ static int vm_playmsgexec(struct ast_channel *chan, const char *data)
+	struct ast_vm_msg_snapshot *msg;
+
+	/* Take a snapshot of the mailbox and walk through each folder's contents */
+	if ((mailbox_snapshot = ast_vm_mailbox_snapshot_create(mailbox, context, NULL, 0, AST_VM_SNAPSHOT_SORT_BY_ID, 0))) {
+		ast_cli(a->fd, VM_STRING_HEADER_FORMAT, "Folder", "Caller ID", "Date", "Duration", "Flag", "ID");
+
+		for (int i=0; i<mailbox_snapshot->folders; i++) {
+			AST_LIST_TRAVERSE(&((mailbox_snapshot)->snapshots[i]), msg, msg) {
+				ast_cli(a->fd, VM_STRING_DATA_FORMAT, msg->folder_name, msg->callerid, msg->origdate, msg->duration,
+						msg->flag, msg->msg_id);
+			}
+		}
+
+		ast_cli(a->fd, "%d Messages Total\n", mailbox_snapshot->total_msg_num);
+		/* done, destroy. */
+		mailbox_snapshot = ast_vm_mailbox_snapshot_destroy(mailbox_snapshot);
+	} else {
+		ast_log(LOG_NOTICE, "Can't create snapshot for voicemail box %s@%s\n", mailbox, context);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int show_messages_for_mailbox(struct ast_cli_args *a)
+{
+	show_mailbox_details(a);
+	ast_cli(a->fd, "\n");
+	show_mailbox_snapshot(a);
+	return 0;
+}
+
+static int forward_message_from_mailbox(struct ast_cli_args *a)
+{
+	const char *from_mailbox = a->argv[2];
+	const char *from_context = a->argv[3];
+	const char *from_folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
+	const char *to_mailbox = a->argv[6];
+	const char *to_context = a->argv[7];
+	const char *to_folder = a->argv[8];
+	return vm_msg_forward(from_mailbox, from_context, from_folder, to_mailbox, to_context, to_folder, 1, id, 0);
+}
+
+static int move_message_from_mailbox(struct ast_cli_args *a)
+{
+	const char *mailbox = a->argv[2];
+	const char *context = a->argv[3];
+	const char *from_folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
+	const char *to_folder = a->argv[6];
+	return vm_msg_move(mailbox, context, 1, from_folder, id, to_folder);
+}
+
+static int remove_message_from_mailbox(struct ast_cli_args *a)
+{
+	const char *mailbox = a->argv[2];
+	const char *context = a->argv[3];
+	const char *folder = a->argv[4];
+	const char *id[] = {a->argv[5]};
+	return vm_msg_remove(mailbox, context, 1, folder, id);
+}
+
+static char *complete_voicemail_show_mailbox(struct ast_cli_args *a)
+{
+	const char *word = a->word;
+	int pos = a->pos;
+	int state = a->n;
+	int which = 0;
+	int wordlen;
+	struct ast_vm_user *vmu;
+	const char *context = "", *mailbox = "";
+	char *ret = NULL;
+
+	/* 0 - voicemail; 1 - show; 2 - mailbox; 3 - <mailbox>; 4 - <context> */
+	if (pos == 3) {
+		wordlen = strlen(word);
+		AST_LIST_LOCK(&users);
+		AST_LIST_TRAVERSE(&users, vmu, list) {
+			if (!strncasecmp(word, vmu->mailbox , wordlen)) {
+				if (mailbox && strcmp(mailbox, vmu->mailbox) && ++which > state) {
+					ret = ast_strdup(vmu->mailbox);
+					AST_LIST_UNLOCK(&users);
+					return ret;
+				}
+				mailbox = vmu->mailbox;
+			}
+		}
+		AST_LIST_UNLOCK(&users);
+	}
+	else if (pos == 4) {
+		/* only display contexts that match the user in pos 3*/
```

Comment:
``<sidenote>``
Personally, I feel like it would be more logical to show all mailboxes in a given context, than all contexts that could match a given mailbox... but I guess this fits with mailbox being required and context defaulting to "default" and thus being optional. Just feels a bit bizarre that it works for me.
``</sidenote>``

Reply:
It does make sense that context is major over user, but people think of the mailbox as user@context and I think it would make the syntax confusing to put them in the other order.  Even if from a sorting perspective it's the other way around.